File: 2021/day13/day13.py
from collections import defaultdict


def read_input():
    with open("input.in", "r") as file:
        dots = defaultdict(int)
        instructions = []
        max_x, max_y = -1, -1
        is_instruction = False
        for line in [line.strip() for line in file.readlines()]:
            if line == "":
                is_instruction = True
                continue
            if not is_instruction:
                x, y = list(map(int, line.split(',')))
                dots[(x, y)] += 1
                max_x = x if max_x < x else y
                max_y = y if max_y < y else y
            else:
                axis, value = line.split()[2].split('=')
                axis = 0 if axis == 'x' else 1
                instructions.append((axis, int(value)))
    return dots, [max_x, max_y], instructions


def fold(dots, max_coordinate, instruction):
    axis, value = instruction
    updated_dots = defaultdict(int)
    if abs(0 - value) > abs(max_coordinate[axis] - value):
        new_max = abs(0 - value) - 1
    else:
        new_max = abs(max_coordinate[axis] - value) - 1
    max_coordinate[axis] = new_max
    for coordinate, dot in dots.items():
        coordinate = list(coordinate)
        if coordinate[axis] != value:
            coordinate[axis] = max_coordinate[axis] - (abs(coordinate[axis] - value) - 1)
            updated_dots[tuple(coordinate)] += dot
    return updated_dots


def print_dots(dots, max_coordinate):
    for j in range(max_coordinate[1] + 1):
        for i in range(max_coordinate[0] + 1):
            if (i, j) not in dots.keys():
                symbol = '.'
            else:
                symbol = '#'
            print(symbol, end=" ")
        print()


def get_number_of_dots(dots):
    return len(dots)


def run_all_instructions(dots, max_coordinate, instructions):
    for instruction in instructions:
        dots = fold(dots, max_coordinate, instruction)
    return dots


def part1():
    dots, max_coordinate, instructions = read_input()
    dots = fold(dots, max_coordinate, instructions[0])
    number_of_dots = get_number_of_dots(dots)
    print(f"Part 1: {number_of_dots}")


def part2():
    dots, max_coordinate, instructions = read_input()
    dots = run_all_instructions(dots, max_coordinate, instructions)
    print("Part 2:")
    print_dots(dots, max_coordinate)


if __name__ == "__main__":
    part1()
    part2()

File: 2021/day13/day13.cpp
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>
#include <map>
#include <sstream>
#include <fstream>
#include <set>

using namespace std;

void part1(const pair<set<pair<int, int>>, vector<pair<char, int>>>& dotsAndInstructions);
void part2(const pair<set<pair<int, int>>, vector<pair<char, int>>>& dotsAndInstructions);
pair<set<pair<int, int>>, vector<pair<char, int>>> readInput();
set<pair<int, int>> fold(set<pair<int, int>> &dots, pair<char, int> &instruction);
void foldAll(set<pair<int, int>> &dots, vector<pair<char, int>> &instructions);
int getNumberOfDots(set<pair<int, int>> &dots);
void printDots(const set<pair<int, int>> &dots);


int main() {
    pair<set<pair<int, int>>, vector<pair<char, int>>> dotsAndInstructions = readInput();
    part1(dotsAndInstructions);
    part2(dotsAndInstructions);
    return 0;
}

void part1(const pair<set<pair<int, int>>, vector<pair<char, int>>>& dotsAndInstructions) {
    set<pair<int, int>> dots = dotsAndInstructions.first;
    vector<pair<char, int>> instructions = dotsAndInstructions.second;
    set<pair<int, int>> updatedDots = fold(dots, instructions[0]);
    int numberOfDots = getNumberOfDots(updatedDots);
    cout << "Part 1: " << numberOfDots << endl;
}

void part2(const pair<set<pair<int, int>>, vector<pair<char, int>>>& dotsAndInstructions) {
    set<pair<int, int>> dots = dotsAndInstructions.first;
    vector<pair<char, int>> instructions = dotsAndInstructions.second;
    foldAll(dots, instructions);
    cout << "Part 2: " << endl;
    printDots(dots);
}

pair<set<pair<int, int>>, vector<pair<char, int>>> readInput() {
    set<pair<int, int>> dots;
    vector<pair<char, int>> instructions;
    string line;
    bool isDot = true;
    ifstream file("input.in", ios::in);

    while (getline(file, line)) {
        if (isspace(line[0])) {
            isDot = false;
        }
        else if (isDot) {
            string x, y;
            stringstream ss(line);
            getline(ss, x, ',');
            getline(ss, y, ',');
            dots.insert({stoi(x), stoi(y)});
        } else {
            string temp, value;
            stringstream ss(line);
            char axis = line[11];
            getline(ss, temp, '=');
            getline(ss, value, '=');
            instructions.emplace_back(axis, stoi(value));
        }
    }
    file.close();

    return {dots, instructions};
}

set<pair<int, int>> fold(set<pair<int, int>> &dots, pair<char, int> &instruction) {
    set<pair<int, int>> updatedDots;
    char axis = instruction.first;
    int value = instruction.second;
    for (auto dot : dots) {
        int x = dot.first, y = dot.second;
        if ((axis == 'x' && x == value) || (axis == 'y' && y == value)) {
            continue;
        }
        if (axis == 'x' && x > value) {
            updatedDots.insert({2 * value - x, y});
        } else if (axis == 'y' && y > value) {
            updatedDots.insert({x, 2 * value - y});
        } else {
            updatedDots.insert({x, y});
        }
    }
    return updatedDots;
}

void foldAll(set<pair<int, int>> &dots, vector<pair<char, int>> &instructions) {
    for (auto instruction : instructions) {
        set<pair<int, int>> updatedDots = fold(dots, instruction);
        dots = updatedDots;
    }
}

int getNumberOfDots(set<pair<int, int>> &dots) {
    return dots.size();
}

void printDots(const set<pair<int, int>> &dots) {
    int max_x = -1, max_y = -1;
    for (auto dot : dots) {
        max_x = (dot.first > max_x) ? dot.first : max_x;
        max_y = (dot.second > max_y) ? dot.second : max_y;
    }
    // Initialize grid
    map<pair<int, int>, char> grid;
    for (int j = 0; j < max_y + 1; j++) {
        for (int i = 0; i < max_x + 1; i++) {
            grid[{i, j}] = '.';
        }
    }
    for (auto dot : dots) {
        grid[{dot.first, dot.second}] = '#';
    }
    for (int j = 0; j < max_y + 1; j++) {
        for (int i = 0; i < max_x + 1; i++) {
            cout << grid[{i, j}] << " ";
        }
        cout << endl;
    }
}
